src/main/java/teammates/ui/webapi/action/GetSessionResultsAsCsvAction.java

```diff
@@ -0,0 +1,90 @@
+package teammates.ui.webapi.action;
+
+import teammates.common.datatransfer.SectionDetail;
+import teammates.common.datatransfer.attributes.FeedbackSessionAttributes;
+import teammates.common.datatransfer.attributes.InstructorAttributes;
+import teammates.common.exception.EntityDoesNotExistException;
+import teammates.common.exception.EntityNotFoundException;
+import teammates.common.exception.ExceedingRangeException;
+import teammates.common.exception.InvalidHttpParameterException;
+import teammates.common.exception.RequestExceedingRangeException;
+import teammates.common.util.Const;
+
+/**
+ * Gets feedback session results in csv.
+ */
+public class GetSessionResultsAsCsvAction extends Action {
+
+    @Override
+    protected AuthType getMinAuthLevel() {
+        return AuthType.PUBLIC;
+    }
+
+    @Override
+    public void checkSpecificAccessControl() {
+        String courseId = getNonNullRequestParamValue(Const.ParamsNames.COURSE_ID);
+        String feedbackSessionName = getNonNullRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_NAME);
+
+        FeedbackSessionAttributes fs = logic.getFeedbackSession(feedbackSessionName, courseId);
+
+        if (fs == null) {
+            throw new EntityNotFoundException(new EntityDoesNotExistException("Feedback session is not found"));
+        }
+
+        InstructorAttributes instructor = logic.getInstructorForGoogleId(courseId, userInfo.id);
+        gateKeeper.verifyAccessible(instructor, fs);
+    }
+
+    @Override
+    public ActionResult execute() {
+        // read in session details
+        String section = getRequestParamValue(Const.ParamsNames.SECTION_NAME);
+        String sectionDetailValue = getRequestParamValue(Const.ParamsNames.SECTION_NAME_DETAIL);
+        SectionDetail sectionDetail = SectionDetail.NOT_APPLICABLE;
+        if (section != null && sectionDetailValue != null && !sectionDetailValue.isEmpty()) {
+            if (!SectionDetail.containsSectionDetail(sectionDetailValue)) {
+                throw new InvalidHttpParameterException("Section detail is invalid.");
+            }
+            sectionDetail = SectionDetail.valueOf(sectionDetailValue);
+        }
+
+        // read in other session-related info
+        String courseId = getNonNullRequestParamValue(Const.ParamsNames.COURSE_ID);
+        String feedbackSessionName = getNonNullRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_NAME);
+        String questionId = getRequestParamValue(Const.ParamsNames.FEEDBACK_QUESTION_ID);
+
+        // read in optional params for csv generation
+        boolean isMissingResponsesShown = getBooleanRequestParamValue(
+                Const.ParamsNames.FEEDBACK_RESULTS_INDICATE_MISSING_RESPONSES);
+        boolean isStatsShown = getBooleanRequestParamValue(Const.ParamsNames.FEEDBACK_RESULTS_SHOWSTATS);
+
+        // read in params for testing
+        String simulateExcessDataForTesting = getRequestParamValue("simulateExcessDataForTesting");
```

Comment:
I am still not entirely convinced by this. I feel that the excess data error should come from the logic layer which should be handled here. 

Why is it part of our API params?

Reply:
The excess data error indeed comes from the logic layer in practice. However for testing it is not practical to generate large amount of data just to trigger the error. Hence, this param is kept there so that the error is triggered manually and exception handling can be tested. But you are right in the way that it does not actually test whether the exception will be thrown at the right time, so it is only tested partially. I'm not really sure if there are better ways to do this.

Reply:
For just testing this on the API layer, you could just mock the logic layer call to throw the relevant exception. For E2E testing, we probably HAVE to generate big data! At least that's what I think :) 

It seems a little odd that we are adding a functionality just for our own internal testing. However, we def have to discuss on this with the rest and see.

Reply:
> For just testing this on the API layer, you can mock the logic layer call to throw the relevant exception.

Sorry but I don't quite get this part. Could you explain a bit more? Thanks :)

Reply:
Adding this line in production code just for testing?

Reply:
Who will throw `ExceedingRangeException`?

Reply:
> Adding this line in production code just for testing?

This is carried forward from https://github.com/TEAMMATES/teammates/pull/8882. This is how the previous developer describes its use:

```
In the `InstructorFeedbackResultsDownloadActionTest` we add a test
where we put in an extra parameter `largeData=true` and then in the Action class
we put an if condition to see if this parameter is not null and true and if it is then we
throw the required exception for the `code` of exception part to be called and assert
in the test if thats what we expect to happen
```

I would like to have some suggestions on improving this if this is not a good programming practice.

> Who will throw `ExceedingRangeException`?

`logic.getFeedbackSessionResultSummaryAsCsv(...)` and `logic.getFeedbackSessionResultSummaryInSectionAsCsv(...)` throw this exception.

Reply:
Then we should meet the condition to let `getFeedbackSessionResultSummaryAsCsv` or `getFeedbackSessionResultSummaryInSectionAsCsv` throw the exception? Or even remove the test because it actually test nothing.

Reply:
> Then we should meet the condition to let `getFeedbackSessionResultSummaryAsCsv` or `getFeedbackSessionResultSummaryInSectionAsCsv` throw the exception? Or even remove the test because it actually test nothing.

The test file is not large enough to trigger the condition that throws this exception. I think it may be better to remove this test then.

Reply:
Hi @niqiukun , sorry missed your earlier question, I was thinking more of something like this:
https://se-education.org/se-book/testing/#stubs

You can just mock the logic layer to throw the relevant exception for testing the API layer. 
For E2E testing, we can just generate a large csv into the database.

Reply:
Introducing mock for logic layer requires a huge effort without help from mocking framework. We might want to do it separately.  I think we can just remove the test till now.